[IAC-1753] Update CIS compliance guide (#426)
gruntwork-io/gruntwork-io.github.io: _posts/2019-10-17-how-to-achieve-cis-benchmark-compliance.adoc

AWS Account or Organization.

To achieve this compliance requirement, enable the IAM Access Analyzer service for every AWS region you have enabled in

> every one of your AWS accounts. Alternatively, you could make use of the link:https://github.com/gruntwork-io/terraform-aws-security/tree/master/modules/iam-access-analyzer-multi-region[`iam-access-analyzer-multi-region`] module

Author: nice 👍

To achieve this compliance requirement, enable the IAM Access Analyzer service for every AWS region you have enabled in
every one of your AWS accounts. Alternatively, you could make use of the `iam-access-analyzer-multi-region` module
available in the Gruntwork Service Catalog, or if you're following this guide, refer to <<apply_account_baseline_root>>, <<apply_account_baseline_logs>> and <<apply_account_baseline_security>> sections of the
deployment walkthrough to see how this module is used from the Landing Zone account baselines and is deployed in each of your AWS accounts.

Once enabled, it will scan only within the boundaries of the AWS Account or Organization it has access to. Only specific

> resources are analyzed and included in the results - e.g. S3 buckets, SQS, etc. (For the full list of resources supported,

Reviewer: @marinalimeira I'm going to remove the link. As you mentioned, it's broken and the only other similar resources I could find are on non-Amazon-controlled websites.

=== Logging
In the Logging section, the Benchmark recommendations target the following services:

* link:https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-user-guide.html[AWS CloudTrail]
* link:https://docs.aws.amazon.com/config/latest/developerguide/WhatIsConfig.html[AWS Config]
* link:https://docs.aws.amazon.com/kms/latest/developerguide/rotate-keys.html[KMS key rotation]

> * link:https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs.html[VPC Flow Logs]

Author: also a nice touch!

Reviewer: Yeah good idea, that makes them consistent!

==== Terraform
This guide uses https://www.terraform.io/[Terraform] to define and manage all the infrastructure as code. If
you're not familiar with Terraform, check out
https://blog.gruntwork.io/a-comprehensive-guide-to-terraform-b3d32832baca[A Comprehensive Guide to Terraform],
https://training.gruntwork.io/p/terraform[A Crash Course on Terraform], and
link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library[How to Use the Gruntwork
Infrastructure as Code Library].

==== Keybase (optional)
As part of this guide, you will create IAM users, including, optionally, credentials for those IAM users. If you
choose to create credentials, those credentials will be encrypted with a PGP key. You could provide the PGP keys
manually, but a more manageable option may be to have your team members to sign up for https://keybase.io[Keybase],
create PGP keys for themselves, and then you can provide their Keybase usernames, and the PGP keys will be retrieved

> automatically.

Reviewer 2: Is it necessarily to include Terragrunt here, as well as any new repositories they need to create, or do you think it's sufficient that we link them to the LZ guide, which will mention those things?

I notice there's terragrunt commands mentioned here, so maybe at least mentioning terragrunt would be a good idea?

Reviewer 2: Ah, I see that the `infrastructure-live` repo is "optional" in the diagram below.. will comment there...

Reviewer: hi @rhoboat, we considered the following options:

1. We include a minor mention of Terragrunt in a paragraph.
2. We basically copy everything that Rho wrote and reword the necessary bits.
3. We do nothing and update the guide later to link to Rho's section once it's published.
4. We do nothing.

In the end, we decided to include a small paragraph (stolen from your work) on Terragrunt. In the future, we could simply de-duplicate work and link to anything you may publish in the LZ guide but I was hesitate to introduce a dependency for the moment.

to store the credentials in an encrypted format.

Use a strong, generated password::
Do NOT re-use passwords from other websites, or any password that you can remember at all. Instead, generate a random,
cryptographically secure, long password (20+ characters). All the password managers mentioned above can generate and
store passwords for you in one step, so use them!

Enable MFA::
Always make sure to
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_mfa_enable.html[enable MFA for your IAM user].
Feel free to use a virtual or hardware MFA device—whichever is easier or required by your company—as either one
dramatically improves the security of your IAM user. Note that using SMS (text messages) for MFA is
https://www.schneier.com/blog/archives/2016/08/nist_is_no_long.html[no longer recommended by NIST] due to known
https://www.theverge.com/2017/9/18/16328172/sms-two-factor-authentication-hack-password-bitcoin[vulnerabilities with the cellular system],
so using a virtual or hardware MFA device is preferable; that said, MFA with SMS is still better than no MFA at all.

> Complete this step before creating any IAM users!

Reviewer 3: This sentence is confusing - which step is it referring to? Could we clarify this a bit?

Author: you're right - it's a bit out of nowhere.... not sure what it was referring to, but now I'd vote let's remove this line for now.

Reviewer: I'm removing for now!

  # older version, so when you upgrade, you should upgrade everyone on your team and your CI servers all at once.
  required_version = "= 0.12.26"
}
----

Next, use the `account-baseline-root` module from the Gruntwork Infrastructure as Code Library:

.infrastructure-modules/landingzone/account-baseline-root/main.tf
[source,hcl]
----
module "root_baseline" {
  source = "git::git@github.com:gruntwork-io/terraform-aws-cis-service-catalog.git//modules/landingzone/account-baseline-root?ref=v0.44.10"

  aws_account_id = var.aws_account_id
  aws_region     = var.aws_region
  name_prefix    = var.name_prefix

  # If you're running the example against an account that doesn't have AWS Organization created, change the following value to true
  create_organization = var.create_organization

  # The child accounts to create
  child_accounts = var.child_accounts

  # IAM users to create in this account
  users = var.users

  # These are variables you only need to set at test time so that everything can be deleted cleanly.
  # You will likely NOT need to set this in any real environments.
  force_destroy_users      = var.force_destroy
  cloudtrail_force_destroy = var.force_destroy

>   config_force_destroy     = var.force_destroy

Reviewer 3: Should we be mentioning them here then, considering this is a production guide?

Author: I think you might be right and we don't want these here!

      is_logs_account = true
    },
    security = {
      email = "root-accounts+security@acme.com"
    },
    shared-services = {
      email = "root-accounts+shared-services@acme.com"
    },
    dev = {
      email = "root-accounts+dev@acme.com"
    },
    stage = {
      email = "root-accounts+stage@acme.com"
    },
    prod = {
      email = "root-accounts+prod@acme.com"
    }
  }

  # The IAM users to create in this account. Since this is the root account, you should only create IAM users for a
  # small handful of trusted admins.
  #
  # NOTE: Make sure to include the IAM user you created manually here! We'll import the user into Terraform state in
  # the next step of this guide, allowing you to manage this user as code going forward.
  users = {
    alice = {
      groups               = ["support"]
      pgp_key              = "keybase:alice"
      create_login_profile = true
      create_access_keys   = false
    },
    bob = {
      groups               = ["billing"]
      pgp_key              = "keybase:bob"
      create_login_profile = true
      create_access_keys   = false
    }
  }
}
----

The example code above does the following:

> . **Create 6 child AWS accounts**. These are the accounts described in the <<child_accounts>> sections.

Reviewer 3: The `<<child_accounts>>` link does not resolve.

Author: Same here: This refers to the Standard Service catalog actually. I wonder if we need to have such a section in our CIS guide too.. https://gruntwork.io/guides/foundations/how-to-configure-production-grade-aws-account-structure/#child_accounts

  "Account": "<SECURITY_ACCOUNT_ID>",
  "Arn": "arn:aws:sts::<SECURITY_ACCOUNT_ID>:assumed-role/OrganizationAccountAccessRole/1597932316055520000"
}
----

You're now ready to deploy the `account-baseline` module in the security account by running `terragrunt apply`:

[source,bash]
----
cd infrastructure-live/security/_global/account-baseline
aws-vault exec security-from-root -- terragrunt apply
----

[.exceptional]
IMPORTANT: On some operating systems, such as MacOS, you may also need to increase your open files limit to avoid "pipe: too many open files" errors by running: `ulimit -n 1024`.

When `apply` finishes, the module will output the encrypted passwords for the users defined above. Send the encrypted
password to each user, along with their user name, and the IAM user sign-in URL for the account. Each user can then
decrypt the password on their own computer (which should have their PGP key) as follows:

[source,bash]
----
echo "<PASSWORD>" | base64 --decode | keybase pgp decrypt
----

=== Apply the security baseline to the other child accounts

Now that your security account is fully configured, you need to apply the security baseline to the remaining child

> accounts (e.g., dev, stage, prod, shared-services) as detailed in the <<child_accounts>> section. Feel free to adjust

Reviewer 3: The `<<child_accounts>>` anchor does not resolve.

Author: This refers to the Standard Service catalog actually. I wonder if we need to have such a section in our CIS guide too.. https://gruntwork.io/guides/foundations/how-to-configure-production-grade-aws-account-structure/#child_accounts

Reviewer: I've reworked the "child accounts" sections.

  link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library[How to Use the Gruntwork
  Infrastructure as Code Library].

Keybase (optional)::
  As part of this guide, you will create IAM users, including, optionally, credentials for those IAM users. If you
  choose to create credentials, those credentials will be encrypted with a PGP key. You could provide the PGP keys
  manually, but a more manageable option may be to have your team members to sign up for https://keybase.io[Keybase],
  create PGP keys for themselves, and then you can provide their Keybase usernames, and the PGP keys will be retrieved
  automatically.

[[gruntwork_solution]]
=== The Gruntwork solution
Gruntwork offers battle-tested infrastructure as code modules to help you create production grade infrastructure in a
fraction of the time it would take to develop from scratch. Each of the code modules are "compliance-ready"; they are
mostly unopinionated by default, but they can be configured for compliance with the right settings.

To further simplify and expedite compliance, the Gruntwork CIS Service Catalog uses both standalone compliance

> modules as well as "wrappers" around the core, unopinionated modules in the Infrastructure as Code Library. The
> standalone compliance modules are designed with the compliance requirements built-in, whereas the wrappers
> call the core modules with configuration values that are compliant with the AWS Foundations Benchmark. You can use both
> types of modules by creating a module of your own (in the case of the wrapper modules, this can be considered a second

Author: NIT: I understand what this says, but I had to read it a few times to grasp it. Might be my English 😛 but it would make it easier if we said something like this:
```suggestion
modules as well as "wrappers" around the core, unopinionated modules in the Infrastructure as Code Library. The
standalone compliance modules are designed with the compliance requirements built-in, whereas the wrappers
call the core IaC modules and pass down configuration values that enforce the AWS Foundations Compliance Benchmark. You can use both
types of modules by creating a module of your own (in the case of the wrapper modules, this can be considered a second
```

Author: or something similar - just to make it explicit which modules are which.

Author: happy to leave it as it is now

wrapper) and using the compliance module as the `source`. Optionally, you can also use `terragrunt` to call your module,
thus creating a chain of IaC modules. For more information on creating your own service catalog of modules please refer
to the link:https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/master/core-concepts.md#create-your-own-service-catalog[Create your own service catalog section]
in the Gruntwork Service Catalog documentation.
 +
 +

.Types of CIS module relationships to avoid repetitive code and minimize the amount of extra work needed to achieve compliance.

> image::/assets/img/guides/cis-compliance/cis-module-relationships.png[]

Author: nice diagram @robmorgan!

Reviewer 2: I'm not sure if users will understand that the `infrastructure-live` repo is optional because this walkthrough is written entirely using that repo. Hmm..

Reviewer 2: I suppose the section below, which talks about how to Create your own service catalog, is sufficient to clarify this point.

Reviewer: @rhoboat ok!

dramatically improves the security of your root user. It is up to your discretion to decide which option is suitable
for your use case. The CIS benchmark recommends using a dedicated or company-owned device for MFA and not a personal
one. This applies to both virtual and hardware devices.

Disable access keys::
Make sure to
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_root-user.html#id_root-user_manage_delete-key[delete the root user's access keys],
so that the only way to login as the root user is via the web console, where MFA is required.

Don't use the root user again::
In the next section, you will create an IAM user in the root account with admin permissions. Once you've created that
IAM user, you should do everything as that IAM user, and more or less never touch the root user account again.
The only time you'll need it is for account recovery situations (e.g., you accidentally deleted the IAM user or lost
your credentials) or for the
https://docs.aws.amazon.com/general/latest/gr/aws_tasks-that-require-root.html[small number of tasks that require root user credentials].

> _This fulfils requirement #1.7 from the CIS 1.3.0 benchmark._

Author: ```suggestion
```

to store the credentials in an encrypted format.

Use a strong, generated password::
Do NOT re-use passwords from other websites, or any password that you can remember at all. Instead, generate a random,
cryptographically secure, long password (20+ characters). All the password managers mentioned above can generate and
store passwords for you in one step, so use them!

Enable MFA::
Always make sure to
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_mfa_enable.html[enable MFA for your IAM user].
Feel free to use a virtual or hardware MFA device—whichever is easier or required by your company—as either one
dramatically improves the security of your IAM user. Note that using SMS (text messages) for MFA is
https://www.schneier.com/blog/archives/2016/08/nist_is_no_long.html[no longer recommended by NIST] due to known
https://www.theverge.com/2017/9/18/16328172/sms-two-factor-authentication-hack-password-bitcoin[vulnerabilities with the cellular system],
so using a virtual or hardware MFA device is preferable; that said, MFA with SMS is still better than no MFA at all.

> wrapper module]. Complete this step before creating any IAM users!

Author: ```suggestion
Complete this step before creating any IAM users!
```

      + force_destroy = true
      + id            = (known after apply)
      + name          = "alice"
      + path          = "/"
      + unique_id     = (known after apply)
    }

# ... (ommitting lots of log output for simplicity) ...

Plan: 160 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
----

This `plan` output is telling you that Terraform will create a bunch of resources, including the `aws_iam_user` named
`alice`. Of course, this user already exists, so we want to `import` the user rather than create it again. The text
next to the `#` gives you the `<ADDRESS>` to use:

----
# module.root_baseline.module.iam_users.aws_iam_user.user["alice"] will be created
----

So the `<ADDRESS>` you want is `module.root_baseline.module.iam_users.aws_iam_user.user["alice"]`. Now, normally, you'd
run `import` right away, but due two Terraform bugs, https://github.com/hashicorp/terraform/issues/13018[#13018] and

> https://github.com/hashicorp/terraform/issues/26211[#26211], `import` doesn't work on certain types of modules—namely,

Author: NIT: It seems like these issues have a fix now. The first one has a fix shown in another issue: https://github.com/hashicorp/terraform/issues/27934 and here https://github.com/hashicorp/terraform/issues/26211#issuecomment-805270266

Author: We might need to test the fix and update the guide.

Author: raise a new issue for this, so resolving comment.

[source,bash]
----
cd infrastructure-live/logs/_global/account-baseline
aws-vault exec logs-from-root -- terragrunt apply
----

[.exceptional]
IMPORTANT: On some operating systems, such as MacOS, you may also need to increase your open files limit to avoid "pipe: too many open files" errors by running: `ulimit -n 1024`.

=== Apply the security baseline to the security account

Now that your logs accounts is fully configured, you need to apply the security baseline to the security account, which
is where all your IAM users and groups will be defined and managed.

Create a new module called `account-baseline-security` in your `infrastructure-modules` repo:

----
infrastructure-modules
  └ landingzone
    └ account-baseline-root
    └ account-baseline-app
    └ account-baseline-security
      └ main.tf
      └ outputs.tf
      └ variables.tf
----

Inside of `main.tf`, configure your AWS provider and Terraform settings:

.infrastructure-modules/landingzone/account-baseline-security/main.tf
[source,hcl]
----
provider "aws" {
  # The AWS region in which all resources will be created
  region = var.aws_region

>   # Require a 2.x version of the AWS provider
>   version = "~> 2.6"

Author: ```suggestion
  # Require a 3.x version of the AWS provider
  version = "~> 3.40"
```

Author: to be consistent with the rest of the AWS provider versions

  source = "git::ssh://git@github.com/gruntwork-io/terraform-aws-cis-service-catalog.git//modules/cleanup-expired-certs?ref=<VERSION>"
}

# Include all settings from the root terragrunt.hcl file
include {
  path = find_in_parent_folders()
}
----

inputs = {
  schedule_expression                = "rate(1 hour)"
  report_cloudwatch_metric           = true
  report_cloudwatch_metric_namespace = "custom/cis"
  report_cloudwatch_metric_name      = "cleanup-expired-iam-certs-count"
}
Just as with the logs and security accounts, you're going to use the `OrganizationAccountAccessRole` IAM role created by
`account-baseline-root` to authenticate to the stage account and all other child accounts. There are many ways to
https://blog.gruntwork.io/a-comprehensive-guide-to-authenticating-to-aws-on-the-command-line-63656a686799[assume an IAM role on the CLI];
for this guide, we're going to keep using `aws-vault`.

Add a new `profile` entry in `~/.aws/config` for your stage account that uses the `root-iam-user` as the
`source_profile`:

[source,text]
----
[profile stage-from-root]
role_arn=arn:aws:iam::<STAGE_ACCOUNT_ID>:role/OrganizationAccountAccessRole
source_profile=root-iam-user
----

> [[iam_access_analyzer_deployment]]

Reviewer 3: This section has been removed but is still being linked to from other parts of the document. We should reconcile that.

Author: Hm, looking at that - it's just that we're no longer deploying that on its own. We're now using the baselines to deploy that. So maybe just remove the refs to this link & make sure the sentences make sense.

Reviewer: @ina-stoyanova @infraredgirl I spoke to Ana and we agreed to resolve this.

  cloudtrail_benchmark_alarm_sns_topic_arn = "arn:aws:sns:${local.aws_region}:${local.accounts.logs}:BenchmarkAlarmTopic"
  ##################################
  # Cross-account IAM role permissions
  ##################################

  # Create groups that allow IAM users in this account to assume roles in your other AWS accounts.
  iam_groups_for_cross_account_access = local.cross_account_groups.cross_account_groups

  # Allow these accounts to have read access to IAM groups and the public SSH keys of users in the group.
  allow_ssh_grunt_access_from_other_account_arns = [
    for name, id in local.accounts :
    "arn:aws:iam::${id}:root" if name != "security"
  ]

  # A list of account root ARNs that should be able to assume the auto deploy role.
  allow_auto_deploy_from_other_account_arns = [
    # External CI/CD systems may use an IAM user in the security account to perform deployments.
    "arn:aws:iam::${local.accounts.security}:root",

    # The shared account contains automation and infrastructure tools, such as CI/CD systems.
    "arn:aws:iam::${local.accounts.shared}:root",
  ]
  auto_deploy_permissions = [
    "iam:GetRole",
    "iam:GetRolePolicy",

>   ]

Reviewer 4: Again, what auto deploy do we do here? Also, you're allowing access from the security account... But this _is_ the security account, so this doesn't quite make sense.

link:https://gruntwork.io/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library/[How to use
the Gruntwork Infrastructure as Code Library] guide.

=== Create the root account

The first step is to create your root account. This account will be the parent of all of your other AWS accounts and
the central place where you manage billing. You create this initial account manually, via a web browser:

. Go to https://aws.amazon.com.

> . Click Create an AWS Account.

Reviewer 5: This button doesn't appear for me. It appears only after "Sign In to the Console"

Reviewer: @marinalimeira This is what I see in the top-right. Are we okay to resolve this one?

<img width="1515" alt="Screen Shot 2021-06-09 at 10 50 03 am" src="https://user-images.githubusercontent.com/178939/121285185-843a3900-c910-11eb-8d14-135e5fa62aba.png">

security standpoint, but also, the `account-baseline-xxx` modules we will use below assume IAM roles, which does not
work with a root user. Later on, we'll create and manage all IAM users as code, but you should create this very first
IAM user manually by
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html#id_users_create_console[following these instructions]:

. Enter a username for your IAM user.
. Select both "programmatic access" and "AWS Management Console access."
. On the next page, click "Attach existing policies to user directly" and attach the `AdministratorAccess` policy.
. Click next a few more times to create the IAM user.
. In a secrets manager, save the IAM sign-in URL, your IAM user's username, the password, and your Access Keys.

[[lock_down_iam_users]]
=== Lock down the root account IAM users

Although IAM users don't have the same powers as a root user, having an IAM user account compromised can still be a
huge problem for your company (especially if that IAM user had admin permissions), so it's still critical to lock down
IAM user accounts as much as possible:

> Use a secrets manager::

Reviewer 5: This section seems duplicated, just a NIT for we to reduce the amount of text in this article

Reviewer: @marinalimeira Overall I agree with you, but as it's only 3 steps I'd rather leave them "as-is" to avoid confusion. Otherwise for "Lock down the root account IAM users", we'd have to say follow the recommendations in "Lock down the root user", but ignore all of them except for:

* Use a secrets manager
* Use a strong, generated password
* Enable MFA

So it ends up getting slightly verbose anyway.

Reviewer: @ina-stoyanova @marinalimeira resolving by leaving it in for now.

to store the credentials in an encrypted format.

Use a strong, generated password::
Do NOT re-use passwords from other websites, or any password that you can remember at all. Instead, generate a random,
cryptographically secure, long password (20+ characters) for the root user. All the password managers mentioned above
can generate and store passwords for you in one step, so use them!

Add security questions to your root account::
CIS recommendation 1.3 suggests that you register your AWS account with security questions. This is so that when someone contacts AWS customer service support from this account they will have to authenticate. To add your security questions to the root account, navigate in the web console for AWS to `My Account` and then the `Personal Information` page.
There you should be able to click on `Configure Security Challenge Questions` and add your questions.

Enable MFA::
Make sure to
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_root-user.html#id_root-user_manage_mfa[enable MFA for your root user].
Feel free to use a virtual or hardware MFA device—whichever is easier or required by your company—as either one

> dramatically improves the security of your root user. It is up to your discretion to decide which option yis suitable for your use case. CIS 1.3 bechmark recommends that any virtual or hardware device you use to fulfil MFA is not personal, but instead dedicated or company-owned.

Author: this now also suggests that hardware or virtual MFA is desired.

cryptographically secure, long password (20+ characters). All the password managers mentioned above can generate and
store passwords for you in one step, so use them!

Enable MFA::
Always make sure to
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_mfa_enable.html[enable MFA for your IAM user].
Feel free to use a virtual or hardware MFA device—whichever is easier or required by your company—as either one
dramatically improves the security of your IAM user. Note that using SMS (text messages) for MFA is
https://www.schneier.com/blog/archives/2016/08/nist_is_no_long.html[no longer recommended by NIST] due to known
https://www.theverge.com/2017/9/18/16328172/sms-two-factor-authentication-hack-password-bitcoin[vulnerabilities with the cellular system],
so using a virtual or hardware MFA device is preferable; that said, MFA with SMS is still better than no MFA at all.
##

wrapper module]. Complete this step before creating any IAM users!
=== Configure the security baseline for the root account

Next, we'll configure a security baseline for the root account that is responsible for creating all the child accounts.

> It will also configure AWS Organizations, IAM Roles, IAM Users, IAM Groups, IAM Password Policies, Amazon GuardDuty, AWS CloudTrail, AWS Config and Security Hub.

Reviewer: Ok great!

We'll be using the `account-baseline-root` module from https://github.com/gruntwork-io/terraform-aws-cis-service-catalog[terraform-aws-cis-service-catalog].

[.exceptional]
IMPORTANT: You must be a [js-subscribe-cta]#Gruntwork subscriber# to access `terraform-aws-service-catalog`.

First, create a _wrapper module_ called `account-baseline-root` in your `infrastructure-modules` repo under the `landingzone` subdirectory:

----
infrastructure-modules
  └ landingzone
    └ account-baseline-root
      └ main.tf
      └ outputs.tf
      └ variables.tf
----

Inside of `main.tf`, configure your AWS provider and Terraform settings:

.infrastructure-modules/landingzone/account-baseline-root/main.tf
[source,hcl]
----
provider "aws" {
  # The AWS region in which all resources will be created
  region = var.aws_region

  # Require a 3.x version of the AWS provider

>   version = "~> 3.40"

Author: I've upped the version of the AWS provider here because that's what we seem to be doing the RefArches with - e.g. Nomisma `lock.hcl` file

Reviewer: Makes sense.

  # Only these AWS Account IDs may be operated on by this template
  allowed_account_ids = [var.aws_account_id]
}

terraform {
  # The configuration for this backend will be filled in by Terragrunt or via a backend.hcl file. See
  # https://www.terraform.io/docs/backends/config.html#partial-configuration
  backend "s3" {}

  # Only allow this Terraform version. Note that if you upgrade to a newer version, Terraform won't allow you to use an
  # older version, so when you upgrade, you should upgrade everyone on your team and your CI servers all at once.

>   required_version = "= 0.14.8"

Author: also upped the TF version here - this is what I've been testing with. Should it be something else? I assume once this repo is also upgraded to TF 0.15, we'll need to update this anyway

Reviewer: Yep, leave it as 0.14.x for now and we can bump to 0.15 afterwards.

}
----

Next, use the `account-baseline-root` module from the Gruntwork Infrastructure as Code Library:

.infrastructure-modules/landingzone/account-baseline-root/main.tf
[source,hcl]
----
module "root_baseline" {
  source = "git::git@github.com:gruntwork-io/terraform-aws-cis-service-catalog.git//modules/landingzone/account-baseline-root?ref=v0.44.10"

  aws_account_id = var.aws_account_id
  aws_region     = var.aws_region
  name_prefix    = var.name_prefix

  # If you're running the example against an account that doesn't have AWS Organization created, change the following value to true
  create_organization = var.create_organization

  # The child accounts to create
  child_accounts = var.child_accounts

  # IAM users to create in this account
  users = var.users

  # These are variables you only need to set at test time so that everything can be deleted cleanly.
  # You will likely NOT need to set this in any real environments.
  force_destroy_users      = var.force_destroy
  cloudtrail_force_destroy = var.force_destroy
  config_force_destroy     = var.force_destroy

>   # Enable IAM Access Analyzer
>   iam_access_analyzer_type   = var.iam_access_analyzer_type
>   iam_access_analyzer_name   = var.iam_access_analyzer_name
>   enable_iam_access_analyzer = var.enable_iam_access_analyzer

Author: This is taken from the `cis-repo/examples/ladningzone/account-baseline-root`.
I believe the IAM Access Analyzer should not be configurable:

```suggestion
  # Enable IAM Access Analyzer
  iam_access_analyzer_type   = var.iam_access_analyzer_type
  iam_access_analyzer_name   = var.iam_access_analyzer_name
```

Reviewer: Yeah agreed, I imagine @marinalimeira probably hasn't updated her example yet.

Author: I'll follow up and leave a comment on the `root` PR

Author: this was already covered in the root PR.

  #     tags   = {
  #       foo = "bar"
  #     }
  #   }
  #
  #   carol = {
  #     groups               = ["user-self-mgmt", "developers", "ssh-users"]
  #     pgp_key              = "keybase:carol_on_keybase"
  #     create_login_profile = true
  #     create_access_keys   = true
  #   }
  # }

  default = {}
}

variable "force_destroy" {
  description = "If set to true, when you run 'terraform destroy', delete all objects from all S3 buckets and any IAM users created by this module so that everything can be destroyed without error. Warning: these objects are not recoverable so only use this if you're absolutely sure you want to permanently delete everything! This is mostly useful when testing."
  type        = bool
  default     = false
}

# ---------------------------------------------------------------------------------------------------------------------
# IAM ACCESS ANALYZER MODULE EXAMPLE PARAMETERS
# ---------------------------------------------------------------------------------------------------------------------

variable "enable_iam_access_analyzer" {
  description = "A feature flag to enable or disable this module."
  type        = bool
  default     = false

> // TODO -> this needs to be true, as IAM Access Analyzer needs to be enabled for CIS across all regions. Maybe we should even remove this variable from the example?

Author: Either remove this variable, or set it at least to `true` by default. See CIS 1.3 requirement #1.21

  type        = string
  default     = "ORGANIZATION"
}

variable "iam_access_analyzer_name" {
  description = "The name of the IAM Access Analyzer module"
  type        = string
  default     = "baseline_root-iam_access_analyzer"
}
----

Finally, add some useful outputs in `outputs.tf`:

.infrastructure-modules/landingzone/account-baseline-root/outputs.tf
[source,hcl]
----
# ---------------------------------------------------------------------------------------------------------------------
# CONFIG OUTPUTS
# ---------------------------------------------------------------------------------------------------------------------

//TODO -> this needs fixing in the example - currently not referencing a real output
//output "config_s3_bucket_name" {
//  description = "The name of the S3 bucket used by AWS Config to store configuration items."
//  value       = module.root_baseline.config_s3_bucket_name
//}
//
//output "config_s3_bucket_arn" {
//  description = "The ARN of the S3 bucket used by AWS Config to store configuration items."
//  value       = module.root_baseline.config_s3_bucket_arn

> //}

Author: These are currently not recognised & thus is invalid tf. This has been taken from `cis-repo/examples/landingzone/account-baseline-root`

Reviewer: Ok, we are probably still waiting to update this example.

Author: Yep, I thought so. I'll leave the related comments in this PR to check back on this.

# ---------------------------------------------------------------------------------------------------------------------
# CLOUDTRAIL OUTPUTS
# ---------------------------------------------------------------------------------------------------------------------

output "cloudtrail_trail_arn" {
  description = "The ARN of the cloudtrail trail."
  value       = module.root_baseline.cloudtrail_trail_arn
}

output "cloudtrail_s3_bucket_name" {
  description = "The name of the S3 bucket where cloudtrail logs are delivered."
  value       = module.root_baseline.cloudtrail_s3_bucket_name
}

output "cloudtrail_s3_bucket_arn" {
  description = "The ARN of the S3 bucket where cloudtrail logs are delivered."
  value       = module.root_baseline.cloudtrail_s3_bucket_arn
}
//todo -> same as last
//output "cloudtrail_kms_key_arn" {
//  description = "The ARN of the KMS key used by the S3 bucket to encrypt cloudtrail logs."
//  value       = module.root_baseline.cloudtrail_kms_key_arn
//}
//
//output "cloudtrail_kms_key_alias_name" {
//  description = "The alias of the KMS key used by the S3 bucket to encrypt cloudtrail logs."
//  value       = module.root_baseline.cloudtrail_kms_key_alias_name

> //}

Author: same here: 

These are currently not recognised & thus is invalid tf. This has been taken from cis-repo/examples/landingzone/account-baseline-root

link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library#manual_tests_terraform[Manual tests for Terraform code] and
link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library#automated_tests_terraform[Automated tests for Terraform code]
for instructions.

Once the module is working the way you want, submit a pull request, get your changes merged into the master branch, and create a new versioned release by using a Git tag. For example, to create a v0.3.0 release:

[source,bash]
----
git add landingzone/account-baseline-root
git commit -m "Add root account baseline wrapper module"
git tag -a "v0.3.0" -m "Created root account baseline module"
git push --follow-tags
----

NOTE: This guide will use https://github.com/gruntwork-io/terragrunt[Terragrunt] and its associated file and folder
structure to deploy Terraform modules. Please note that *Terragrunt is NOT required for using Terraform modules from
the Gruntwork Infrastructure as Code Library.* Check out link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library[How to use the Gruntwork Infrastructure as Code Library]
for instructions on alternative options, such as how to

> link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library#deploy_using_plain_terraform[deploying how to use plain terraform].

Author: is this what we want for this guide too? None of our examples are currently with terragrunt.

  # connectivity issues, we limit the parallelism here.
  extra_arguments "parallelism" {
    commands  = get_terraform_commands_that_need_parallelism()
    arguments = ["-parallelism=2"]
  }
}
----

[.exceptional]
IMPORTANT: We **strongly** recommend setting Terraform parallelism to a low value (e.g., `-parallelism=2`), as shown above, with the `account-baseline-xxx` modules. This is because these modules deploy multi-region resources (e.g., GuardDuty, AWS Config, etc), and for each region, Terraform spins up a separate process, so if you don't limit the parallelism, it may peg all your CPU cores and lead to network connectivity errors.

Set the variables for the `account-baseline-root` module in this environment in the `inputs = { ... }` block of `terragrunt.hcl`:

.infrastructure-live/root/_global/account-baseline/terragrunt.hcl
[source,hcl]
----
inputs = {
  # Fill in your region you want to use (only used for API calls) and the ID of your root AWS account
  aws_region     = "us-east-2"
  aws_account_id = "<ROOT_ACCOUNT_ID>"

  # Prefix all resources with this name
  name_prefix = "<COMPANY_NAME>-root"

>   # If you've already created an AWS Organization in your root account, you'll be able to import it later in this guide

Author: ```suggestion
  # If you've already created an AWS Organization in your root account, set this to false and you'll be able to import it later in this guide.
```

https://www.schneier.com/blog/archives/2016/08/nist_is_no_long.html[no longer recommended by NIST] due to known
https://www.theverge.com/2017/9/18/16328172/sms-two-factor-authentication-hack-password-bitcoin[vulnerabilities with the cellular system],
so using a virtual or hardware MFA device is preferable; that said, MFA with SMS is still better than no MFA at all.
##

wrapper module]. Complete this step before creating any IAM users!
=== Configure the security baseline for the root account

Next, we'll configure a security baseline for the root account that is responsible for creating all the child accounts.
It will also configure AWS Organizations, IAM Roles, IAM Users, IAM Groups, IAM Password Policies, Amazon GuardDuty, AWS CloudTrail, AWS Config and Security Hub.

> We'll be using the `account-baseline-root` module from https://github.com/gruntwork-io/terraform-aws-service-catalog[terraform-aws-service-catalog].

Reviewer: I think the money shot here is actually using the `account-baseline-root` module from the CIS service catalog instead. I'll go ahead and change it!

Reviewer: Actually don't worry, I can see you've changed it in a subsequent commit.

.infrastructure-live/root/_global/account-baseline/terragrunt.hcl
[source,hcl]
----
locals {
  accounts = {
    root = "216044045972"

>   }

Reviewer 2: NIT: Perhaps this ID could be stored as a `root` key in the `accounts.json` that I mention in the comment under the `logs` section below.

  # Both buckets will created in the logs account by account-baseline-root
  config_s3_bucket_name     = "acme-config-bucket-logs"
  cloudtrail_s3_bucket_name = "acme-cloudtrail-logs"
}

inputs = {
  # Fill in your region you want to use (only used for API calls) and the ID of your root AWS account

>   aws_region     = "us-east-1"

Reviewer 2: NIT: This could be `local.aws_region`, I think.

.infrastructure-live/logs/_global/account-baseline/terragrunt.hcl
[source,hcl]
----
locals {
  accounts = {
    logs     = "409800740445"
    security = "673123100581"
    shared   = "384759303421"
    dev      = "293847503945"
    stage    = "384924092834"
    prod     = "784260063686"
    root     = "216044045972"

>   }

Reviewer 2: So, in the LZ guide, I've moved these into the root `accounts.json` that gets included in the root `common.hcl` as `accounts = jsondecode(file("accounts.json"))`. Then every `terragrunt.hcl` that needs it does

```
locals {
  # A local for more convenient access to the accounts map.
  accounts = local.common_vars.locals.accounts
  ...
}
```

It's a nit, but wonder if it's something you want to follow here.

  # Both buckets are created in the logs account by account-baseline-root
  config_s3_bucket_name     = "acme-config-bucket-logs"
  cloudtrail_s3_bucket_name = "acme-cloudtrail-logs"

  # The Cloudtrail KMS Key is deployed at the logs account but it's value is an output from the root account.

>   cloudtrail_kms_key_arn = "arn:aws:kms:us-east-1:409800740445:alias/cloudtrail-acme"

Reviewer 2: NIT: Same nit as below on replacing hardcoded values with placeholders.

  # A local for convenient access to the security account root ARN.
  security_account_root_arn = "arn:aws:iam::${local.accounts.security}:root"
}

inputs = {
  # Fill in your region you want to use (only used for API calls) and the ID of your root AWS account

>   aws_region     = "us-east-1"

Reviewer 2: NIT: This could be `local.aws_region`, I think.

>   aws_account_id = local.accounts["logs"]

Reviewer: Nice 👍

Reviewer 2: NIT: I believe this could simply be `local.accounts.logs`, as well.

  # Prefix all resources with this name
  name_prefix = "acme-logs"
 ################################
  # Parameters for AWS Config
  ################################
  # Send Config logs to the common S3 bucket.
  config_s3_bucket_name = local.config_s3_bucket_name

  # Send Config logs and events to the logs account.
  config_central_account_id = local.accounts.logs

  # Do not allow objects in the Config S3 bucket to be forcefully removed during destroy operations.
  config_force_destroy = false

  #  This is the Logs account, so we create the SNS topic for aggregating Config logs from all accounts.
  config_should_create_sns_topic = true

  # All of the other accounts send logs to this account.
  config_linked_accounts = [
  for name, id in local.accounts :

>   id if name != "logs"

Reviewer 2: NIT: In my view, this is a line under the for loop, so it seems like it should be further indented. (We need to update the ref arch generation to reflect this too.)

  ]

  ################################
  # Parameters for CloudTrail
  ################################

  # Send CloudTrail logs to the common S3 bucket.
  cloudtrail_s3_bucket_name = local.cloudtrail_s3_bucket_name

  # All of the other accounts send logs to this account.
  cloudtrail_allow_kms_describe_key_to_external_aws_accounts = true
  cloudtrail_external_aws_account_ids_with_write_access = [
  for name, id in local.accounts :

>   id if name != "logs"

Reviewer 2: NIT: In my view, this is a line under the for loop, so it seems like it should be further indented. (We need to update the ref arch generation to reflect this too.)

.infrastructure-live/security/_global/account-baseline/terragrunt.hcl
[source,hcl]
----
locals {
  accounts = {
    logs     = "409800740445"
    security = "673123100581"
    shared   = "384759303421"
    dev      = "293847503945"
    stage    = "384924092834"
    prod     = "784260063686"
    root     = "216044045972"
  }

  # Both buckets are created in the logs account by account-baseline-root
  config_s3_bucket_name     = "acme-config-bucket-logs"
  cloudtrail_s3_bucket_name = "acme-cloudtrail-logs"

  # The Cloudtrail KMS Key is deployed at the logs account but it's value is an output from the root account.

>   cloudtrail_kms_key_arn = "arn:aws:kms:us-east-1:216044045972:alias/cloudtrail-acme"

Reviewer 2: NIT: Is it worth replacing these hardcoded values with placeholder values, with a comment above them telling them how to get them? Like:

```
  # Use the S3 bucket and KMS key that were already created in the logs account by account-baseline-root
  cloudtrail_s3_bucket_name = "<CLOUDTRAIL_BUCKET_NAME>"
  cloudtrail_kms_key_arn    = "<CLOUDTRAIL_KMS_KEY_ARN>"
```

Reviewer 2: I think what you have currently works fine. In either case, it's up to the customer whether they want to store the values in locals, and if you have them set it there, then it makes no difference that you reference the local here.

Reviewer 2: I made this modification in #442, so we store the bucket names and arns in `common.hcl`. Updating the LZ guide, too.

  # A local for convenient access to the security account root ARN.
  security_account_root_arn = "arn:aws:iam::${local.accounts.security}:root"
}
input = {
 # Fill in your region you want to use (only used for API calls) and the ID of your root AWS account

>   aws_region     = "us-east-1"

Reviewer 2: NIT: This could be `local.aws_region`, I think.

Reviewer 5: This file is outdated, I updated this earlier this morning :D

Reviewer 5: So this is ✅

Reviewer: I'll resolve this one.

>   aws_account_id = local.accounts["security"]

Reviewer 2: NIT: This could be `local.accounts.security`.

for each of those child accounts—including enabling MFA and deleting the root user's access keys—and (almost) never use
those root users again.

> [[apply_account_baseline_logs]]

Author: not just here, but nicely done on the consistent formatting 👍 😄

.infrastructure-live/stage/_global/account-baseline/terragrunt.hcl
[source,hcl]
----
locals {
  aws_region = "us-east-1"

  accounts = {
    logs     = "409800740445"
    security = "673123100581"
    shared   = "384759303421"
    dev      = "293847503945"
    stage    = "384924092834"
    prod     = "784260063686"
    root     = "216044045972"

>   }

Reviewer 2: NIT: See comment in `logs` section about DRYing this up.

  # Both buckets are created in the logs account by account-baseline-root
  config_s3_bucket_name     = "acme-config-bucket-logs"
  cloudtrail_s3_bucket_name = "acme-cloudtrail-logs"

  # The Cloudtrail KMS Key is deployed at the logs account but it's value is an output from the root account.

>   cloudtrail_kms_key_arn = "arn:aws:kms:us-east-1:409800740445:alias/cloudtrail-acme"

Reviewer 2: NIT: Same nit about hardcoded values.

  # A local for convenient access to the security account root ARN.
  security_account_root_arn = "arn:aws:iam::${local.accounts.security}:root"
}

inputs = {
  # Fill in your region you want to use (only used for API calls) and the ID of your root AWS account
  aws_region     = local.aws_region

>   aws_account_id = local.accounts["stage"]

Reviewer 2: NIT: This could be `local.accounts.stage`.

  # Prefix all resources with this name

>   name_prefix = "stage-logs"

Reviewer 2: Hm, this one I'm not sure about. Setting the `name_prefix` for the environment + account is not what we're currently doing in the ref arch deploys. See the [LZ guide](https://gruntwork.io/guides/foundations/how-to-configure-production-grade-aws-account-structure/#prepare-your-infrastructure-live-repository) for how we set `name_prefix` (in `common.hcl`, usually as the company name, and then in the root `terragrunt.hcl`, so that no downstream `terragrunt.hcl`s need to set it). It makes me wonder if our ref arch is setting the `name_prefix` a little too generic? Pinging @bwhaley for another opinion.

Reviewer 6: It does need to be something unique as it will ultimately be used to create the S3 bucket, which is a global namespace among all AWS users. Identifying a good rule of thumb is tricky because it must be both (a) short enough to not run in to length constraints, and (b) globally unique. `<COMPANY_NAME>` is usually unique enough, but runs the risk of being too long. Many customers have chosen a prefix that is an acronym or shortened identifiable version of the company name, like `phc` or `mf`. 

I think we ought to share this context in the comment on the preceding line, and go with something like:

```
name_prefix = "<SOME UNIQUE IDENTIFIER>-logs"
```

Reviewer 2: Thanks Ben. Seems like we need to update the LZ guide and the Ref Arch code, too. Looking at the `terraform-aws-architecture-catalog`, we set `name_prefix` is only [one place](https://github.com/gruntwork-io/terraform-aws-architecture-catalog/blob/b1b7049d9d510d12d0c79ab70b0d0acf055a526c/templates/infrastructure-live/_root/common.hcl#L13-L14), which gets passed around from the [root terragrunt](https://github.com/gruntwork-io/terraform-aws-architecture-catalog/blob/b1b7049d9d510d12d0c79ab70b0d0acf055a526c/templates/infrastructure-live/_root/terragrunt.hcl#L25) (and actually it's [duplicated here](https://github.com/gruntwork-io/terraform-aws-architecture-catalog/blob/b1b7049d9d510d12d0c79ab70b0d0acf055a526c/templates/infrastructure-live/_root/terragrunt.hcl#L73), which we can simplify). That means it's not unique for the S3 buckets, right?

Reviewer 2: I feel like I must be missing something. It can't be that we use the same `name_prefix` everywhere. Maybe it does get overridden in each account's terragrunt, but I'm not seeing it.

Reviewer 2: @bwhaley Thoughts on this?

  allow_auto_deploy_from_other_account_arns = [
    # External CI/CD systems may use an IAM user in the security account to perform deployments.
    local.security_account_root_arn,

    # The shared account contains automation and infrastructure tools, such as CI/CD systems.
    "arn:aws:iam::${local.accounts.shared}:root",
  ]

  # Assuming the auto-deploy role will grant access to these services.
  auto_deploy_permissions = [
    "iam:GetRole",
    "iam:GetRolePolicy",
  ]

  # Configures the auto deploy max session duration to be 4 hours.
  max_session_duration_machine_users = 14400

  # Configures the max session duration for roles that humans use to be 8 hours.
  max_session_duration_human_users = 28800

  service_linked_roles = ["autoscaling.amazonaws.com"]
  ##################################
  # KMS grants
  ##################################

  # These grants allow the autoscaling service-linked role to access to the AMI encryption key so that it
  # can launch instances from AMIs that were shared from the shared-services account.
  kms_grant_regions = {
    ami_encryption_key = local.aws_region
  }
  kms_grants = {
    ami_encryption_key = {

>       kms_cmk_arn       = "arn:aws:kms:us-east-1:${local.accounts.shared}:alias/ami-encryption"

Reviewer 2: NIT: I might remove the reference to `us-east-1` here, using `local.aws_region` or some such, instead.
Or maybe this whole arn should be a placeholder text.

  # Set to point to the AWS Account that has set up the Security Hub organisation and has sent out invitations to other AWS Child accounts. For example, get the `root` account to send invitations to `security`, `logs`, `dev`, etc.
  security_hub_associate_to_master_account_id = "<ROOT_ACCOUNT_ID>"
  security_hub_external_member_accounts = { security = {
                                               account_id = "<LOGS_ACCOUNT_ID>"
                                               email      = "root-accounts+logs@acme.com"
                                             }

>                                         }

Author: I'm wondering why this is removed? Is it no longer necessary?

Reviewer 5: Because this is set in the root account. This account only needs to add the root account id.

Reviewer: @ina-stoyanova @marinalimeira let me know if we can resolve this one?

link:https://github.com/gruntwork-io/cloud-nuke[`cloud-nuke defaults-aws`] command. It will remove the default VPC from
all regions in an account, saving you the hassle of creating flow logs in each default VPC.

> [[configure_monitoring]]

Author: I feel like in this section we should also include the IAM Access Analyzer & the Security Hub too. My thinking is that they're also monitoring tools for any security issues. Curious to get more thoughts on this :)

1.1,<<security_questions>>,Complete the contact details on the AWS account page
1.2,<<security_questions>>,Complete the security contact information on the AWS account page
1.3,<<security_questions>>,Answer the security questions on the AWS account page

> 1.4,<<configure_security_hub>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure that no access key exists for the root user

Author: **_Context:_** Technically, I believe Security Hub isn't part of the CIS official requirements. It is, however, a useful tool to monitor that for example this is a fulfilled rule. It checks currently against CIS 1.2 benchmark, not even 1.3.

Reviewer 3: True. If there's a better section to link to here, let me know.

Author: I think by adding the section for Security hub, this makes more sense now

Reviewer: Yeah, it looks like it's required in CIS 1.4.0, so we'll probably be adding it in the future anyway.

> 1.4,<<configure_security_hub>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure that no access key exists for the root user

Author: ```suggestion
1.4,<<configure_security_hub>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure you are notified if the root user has access keys set.
```

1.5,<<root_mfa>>,Manually configure MFA for the root user
1.6,<<root_mfa>>,Use a Yubikey (or other hardware MFA) for the root user
1.7,<<manual_steps>>,Take manual steps to complete this recommendation

> 1.8-9,<<apply_account_baseline_security>>,Use the `account-baseline-security` module to create users

Author: ```suggestion
```

1.10,<<configure_authentication>>,Configure authentication using SAML or IAM
1.11,<<apply_account_baseline_security>>,Use the `account-baseline-security` module to create users

> 1.12,<<configure_security_hub>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure that there are no unused credentials

Author: ```suggestion
1.12,<<apply_account_baseline_security>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure that there are no unused credentials
```

Author: I believe the Security Hub monitors this, but we have an AWS Config rule for this that should get kicked in

Reviewer 3: This is a bit tricky, since those are links pointing to the specific sections in the guide, so `apply_account_baseline_all` won't work. I guess I could put multiple links pointing to each account baseline module

> 1.13,<<configure_security_hub>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure that there are no extra access keys

Author: Same as above:

```suggestion
1.13,<<apply_account_baseline>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure that there are no extra access keys
```

> 1.13,<<configure_security_hub>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure that there are no extra access keys

Author: Same as above.

> 1.14,<<configure_security_hub>>,Use the Gruntwork Security Hub module to enable AWS Security Hub to ensure that there are no unused access keys

Author: Same as above.

> 1.15,<<apply_account_baseline_security>>,Use the `account-baseline-security` module to create users and groups

Author: 👍

> 1.16,<<apply_account_baseline_security>>,Use the `account-baseline-security` module to create best-practices groups and roles

Author: 👍

Author: We could be more explicit here and say something like "to ensure no full-access policies are attached to any groups or users"

> 1.17,<<apply_account_baseline_security>>,Use the `account-baseline-security` module to create a support group

Author: 👍

1.18,<<iam_roles_for_instances>>,Use Gruntwork modules to ensure EC2 instances use roles for access
1.19,<<cleanup_expired_certs>>,Use Gruntwork modules to automatically remove expired certificates from IAM
1.20,<<s3_buckets_deployment>>,Use the `private-s3-bucket` module
1.21,<<iam_access_analyzer>>,Use Gruntwork modules to enable IAM Access Analyzer across regions
1.22,<<account_structure>>,Use the security AWS account as described in the Gruntwork production-grade AWS account structure
2.1.1-2.1.2,<<s3_buckets_deployment>>,Use the `private-s3-bucket` module
2.2.1,<<configure_ebs_encryption>>,Use Gruntwork modules to configure AWS EBS encryption
3.1-3.4,<<apply_account_baseline_logs>>,Use the `account-baseline-*` modules to ensure Cloudtrail is enabled and configured in all regions

> 3.5,<<apply_account_baseline_security>>,Use the `account-baseline-security` module to ensure AWS Config is enabled

Author: ```suggestion
3.5,<<apply_account_baseline_security>>,Use the `account-baseline-security` module to ensure AWS Config is enabled in all regions
```

> 3.6-3.7,<<apply_account_baseline_logs>>,Use the `account-baseline-*` modules to ensure Cloudtrail is enabled and configured in all regions

Author: ```suggestion
3.6<<apply_account_baseline_logs>>,Use the `account-baseline-*` modules to ensure Cloudtrail S3 bucket has enabled access logging```

3.8,<<kms>>,Use the KMS module
3.9,<<vpc_flow_logs>>,Use the Gruntwork CIS-compliant `vpc` service to provision VPCs with flow logs enabled

> 3.10-3.11,<<apply_account_baseline_logs>>,Use the `account-baseline-*` modules to ensure Cloudtrail is enabled and configured in all regions

Author: ```suggestion
3.10-3.11,<<apply_account_baseline_logs>>,Use the `account-baseline-*` modules to ensure Object-level logging is enabled for S3 buckets for read and write events
```

> 4.1-4.15,<<maintain_compliance_monitoring>>,The CloudWatch Logs metrics filters wrapper module will satisfy each recommendation

Author: 👍

5.1,<<maintain_compliance_networking>>,Use the Gruntwork CIS-compliant `vpc` service to ensure there is no public remote access
5.2,<<maintain_compliance_networking>>,Use the Gruntwork CIS-compliant `vpc` service for a secure network configuration
5.3,<<maintain_compliance_networking>>,Use the `cloud-nuke` tool to remove all default security groups

> 5.4,<<maintain_compliance_networking>>,Use the Gruntwork CIS-compliant `vpc` service to configure least-privilege routing by default

Author: nice
